source/pki_utils.c

```diff
@@ -1,298 +1,816 @@
+    if (status == errSecDuplicateItem) {
+        struct aws_array_list cert_chain_list;
+
+        if (aws_array_list_init_dynamic(&cert_chain_list, alloc, 2, sizeof(struct aws_byte_buf))) {
+            return AWS_OP_ERR;
+        }
+
+        if (aws_decode_pem_to_buffer_list(alloc, public_cert_chain, &cert_chain_list)) {
+            aws_array_list_clean_up(&cert_chain_list);
+            return AWS_OP_ERR;
+        }
+
+        struct aws_byte_buf *root_cert_ptr = NULL;
+        aws_array_list_get_at_ptr(&cert_chain_list, (void **)&root_cert_ptr, 0);
+        assert(root_cert_ptr);
+        CFDataRef root_cert_data = CFDataCreate(cf_alloc, root_cert_ptr->buffer, root_cert_ptr->len);
+
+        if (root_cert_data) {
+            certificate_ref = SecCertificateCreateWithData(cf_alloc, root_cert_data);
+            CFRelease(root_cert_data);
+        }
+
+        aws_cert_chain_clean_up(&cert_chain_list);
+        aws_array_list_clean_up(&cert_chain_list);
+    } else {
+        certificate_ref = (SecCertificateRef)CFArrayGetValueAtIndex(import_output, 0);
+        /* SecCertificateCreateWithData returns an object with +1 retain, so we need to match that behavior here */
+        CFRetain(certificate_ref);
+    }
+
+    if (certificate_ref) {
+        SecIdentityRef identity_output;
+        bool cleanup_import_output = status != errSecDuplicateItem;
+        status = SecIdentityCreateWithCertificate(import_keychain, certificate_ref, &identity_output);
+
+        CFRelease(certificate_ref);
+        CFRelease(import_keychain);
+        if (import_output && cleanup_import_output) {
+            CFRelease(import_output);
+        }
+
+        if (status == errSecSuccess) {
+            CFTypeRef certs[] = {identity_output};
+            *identity = CFArrayCreate(cf_alloc, (const void **)certs, 1L, &kCFTypeArrayCallBacks);
+            return AWS_OP_SUCCESS;
+        }
+    } else {
+        CFRelease(import_keychain);
+    }
+
+    return AWS_OP_ERR;
+}
+
+int aws_import_pkcs12_to_identity(
+    CFAllocatorRef cf_alloc,
+    struct aws_byte_buf *pkcs12_buffer,
+    struct aws_byte_buf *password,
+    CFArrayRef *identity) {
+    CFDataRef pkcs12_data = CFDataCreate(cf_alloc, pkcs12_buffer->buffer, pkcs12_buffer->len);
+    CFArrayRef items = NULL;
+
+    CFMutableDictionaryRef dictionary = CFDictionaryCreateMutable(cf_alloc, 0, NULL, NULL);
+
+    CFStringRef password_ref = CFSTR("");
+
+    if (password->len) {
+        password_ref = CFStringCreateWithBytes(cf_alloc, password->buffer, password->len, kCFStringEncodingUTF8, false);
+    }
+
+    CFDictionaryAddValue(dictionary, kSecImportExportPassphrase, password_ref);
+
+    OSStatus status = SecPKCS12Import(pkcs12_data, dictionary, &items);
+    CFRelease(pkcs12_data);
+
+    if (password_ref) {
+        CFRelease(password_ref);
+    }
+
+    CFRelease(dictionary);
+
+    if (status == errSecSuccess) {
+        CFTypeRef item = (CFTypeRef)CFArrayGetValueAtIndex(items, 0);

```

Comment:
assert(items)?

Reply:
we already checked that the operation was successful.